### DISCUSSION TITLE
The badge does not work on latest Samsung

### DISCUSSION THREAD

**Discussion by happymago** [robingenz/capacitor-badge#26] [General]
**Plugin version:**

isSupported() method returns false. When digging into it more, I have found the shortcutBadger this plugin uses is not maintained anymore. And there is still this open issue https://github.com/leolin310148/ShortcutBadger/issues/266. It seems Android has come up with its own supported badge/notification since android 8. I wonder how everyone else is using this plugin?

**Platform(s):**

latest samsung
latest pixel

**Current behavior:**

isSupported return false

**Expected behavior:**

isSupported return true

**Steps to reproduce:**

**Related code:**

```
insert short code snippets here
```

**Other information:**

**Capacitor doctor:**

```
insert the output from `npx cap doctor` here
```

**Comment by robingenz** [comment]
Hi @happymago,
thank you for your request.
I'll look at using Android's own badge setting interface to replace or extend ShortcutBadger once I have a little more time.
I will get back to you.

**Comment by drtomato** [comment]
Hi @robingenz,

I use your plugin on IOS, where it works excellently! However, I am also migrating to Android and unfortunately here it doesn't work. I have tried to run both my app as well as the Capacitor plugin demo on an upgraded Samsung Galaxy S4 running LineageOS and a Samsung Galaxy S10e running Android 11. It seems that the badge settings are picked up properly inside the plugin, the problem is that the actual badge never appears on the screen. As happymago commented above, the problem seems to be inside ShortcutBadger. I can add more info if needed but if you plan to replace ShortcutBadger with Android's own interface maybe that isn't necessary. Anyway, your efforts are appreciated :)

**Comment by robingenz** [comment]
So I've been looking into this. 
First of all: I have two Android devices to test here: A Samsung Galaxy S4 (Android 5) and a OnePlus 7T (Android 11).

On the Samsung Galaxy S4 this plugin works, because this Samsung OS version is supported by ShortcutBadger. On the OnePlus 7T, the plugin does not work. 

Android supports so-called "Notification Badges" since Android 8.0 (see [here](https://developer.android.com/training/notify-user/badges)). These appear when an app has an active notification. I could not find any other way to display badges. If I have missed a possibility and someone has experience with it, I would be very grateful for any advice.
Since these badges are thus dependent on the notifications of an app and Capacitor already offers a very good official plugin for this, I recommend using the Local Notification plugin if ShortcutBadger does not support the device.

Here is some sample code:

```ts
import { LocalNotifications } from '@capacitor/local-notifications';
import { Badge } from '@robingenz/capacitor-badge';

const set = async (count: number) => {
  const { isSupported } = await Badge.isSupported();
    if (isSupported) {
      await Badge.set({ count });
    } else {
      await LocalNotifications.schedule({
        notifications: [
          {
            id: 1,
            title: 'My Title',
            body: 'My Body',
          },
        ],
      });
    }
};
```

From my point of view, it is outside the scope of this plugin to support local notifications on Android. **App icon badges are simply not a standalone feature on Android but a side effect of notifications.**

I hope you can understand my point of view. I will turn this issue into a discussion so that we can gather other opinions and suggestions there as well.

*BTW*: On Samsung you can switch between Badges with Numbers or Dot-style Badges (see [here](https://www.samsung.com/hk_en/support/mobile-devices/android-o-os-app-icon-can-show-badges-with-numbers-or-dot-style-badges/))? Has anyone tried this?

**Reply by drtomato** [reply]
Thanks for your research, this is interesting. I agree completely with you that LocalNotifications is outside the scope of your plugin. The app I am working on is actually using LocalNotifications but unfortunately this has not helped - neither in the Pixel 3a API 29 and Pixel 4 API 30 simulators, nor the Galaxy S4 (LineageOS, Android 11) and Galaxy S10e (Samsung, Android 11) has this made badges appear, with one very strange exception - on the S10e, the app icon badge was shown at one instance after I reinstalled the app but after having opened the app and put it in the background, the badge was gone and has not re-appeared since, even as I tried to reproduce the sequence of events. Maybe the presence of the plugin (ShortcutBadger really) interferes with Android's internal logic somehow so I will next make an experiment by removing the plugin completely on Android and see what happens. I have made an attempt to switch the badge styles but on the emulators, this made them crash and on my physical phones, the settings look different and I haven't been able to find what to change (at the moment, I will give priority to just make any badge appear and deal with looks later :) ).

**Reply by robingenz** [reply]
Unfortunately, I have only been able to test the Local Notification Plugin with the badge on my OnePlus 7T so far. There it has worked. The next time I have access to another Android device (e.g. Pixel 4), I will test it there as well. I might have to make a few adjustments to the Local Notification Plugin for this. But that should not be a problem.

> Maybe the presence of the plugin (ShortcutBadger really) interferes with Android's internal logic somehow

At least I wouldn't know about it. Please let me know if you have any new information.

**Reply by drtomato** [reply]
Latest update:

- I deleted the Pixel3a with Android 10 (API 29) since Google Play updates now have to be at least Android 11 (API 30)
- I am now testing on the emulators Pixel 4 with Android 11 (API 30) and Pixel 5 with Android 12 (API 31) and the physical devices Samsung Galaxy S4 with LineageOS Android 11 (API 30) and Samsung Galaxy S10e with Samsung Android 12 (API 31)
- I deleted all previous installations of my app and rebuilt and reinstalled from scratch, keeping the pre-existing dependencies and usages of LocalNotifications and Badge. I use LocalNotifications and Badge separately; LocalNotifications for geo- and time-based notifications and Badge for indicating the number of messages in the app's inbox (which is unrelated to the notifications)
- On the emulators, the dotted app badges now work! They are related to the number of open notifications and not affected by the badge number setting (which is the opposite of what I want...)
- On the Galaxy S4, neither dotted nor numeric app badges work, irrespective of the number of open notifications or the badge number setting
- On the Galaxy S10e, numeric app badges now work (!) but only as long as there is at least one open notification (!!). So this almost works as I want it to, except for the fact that of course I want the badge to be independent of the notifications...
- One conclusion is that it is important to delete the app before installing a new version in order for plugin config changes to apply.

**Reply by robingenz** [reply]
Hi @drtomato,
Thanks for the updates. This is really helpful, feel free to continue. Everyone else is invited to do so as well.

> On the emulators, the dotted app badges now work! They are related to the number of open notifications and not affected by the badge number setting (which is the opposite of what I want...)

Does the badge plugin have any effect at all or does only the notification plugin work? What is the return of `Badge.isSupported()`?

> On the Galaxy S4, neither dotted nor numeric app badges work, irrespective of the number of open notifications or the badge number setting

Do you have a special Android version installed?
With my Samsung Galaxy S4 it works without problems (Android 5).
What is the return of `Badge.isSupported()`?

> On the Galaxy S10e, numeric app badges now work (!) but only as long as there is at least one open notification (!!). So this almost works as I want it to, except for the fact that of course I want the badge to be independent of the notifications...

Yes, of course, that should be independent.

**Reply by drtomato** [reply]
Hi @robingenz 

- On the emulators (both Pixel 4 and Pixel 5) as well as on the Galaxy S4
In `ShortcutBadger.isBadgeCounterSupported`:
Call to `BroadcastHelper.sendIntentExplicitly` ==> me.leolin.shortcutbadger.ShortcutBadgeException: unable to resolve intent: Intent { act=me.leolin.shortcutbadger.BADGE_COUNT_UPDATE (has extras) }
which results in the log message:
Badge counter seems not supported for this platform: unable to resolve intent: Intent { act=android.intent.action.BADGE_COUNT_UPDATE (has extras) }

...so I guess it is only to be expected that badges don't work. 

- I don't use the stock Android on the Galaxy S4 since it was outdated and I needed a newer API version for my app to run. So I updated it to LineageOS Android 11 (https://lineageos.org) in order to have something to test on. There seems to be some glitches with other features on LineageOS as well so I think this platform can be ignored for the sake of this discussion. I don't care much about 100% compatibility with LineageOS anyway, it was the only way I found to run a modern Android version on my old device.

- The most well-behaved platform, from my perspective, is the Samsung Galaxy S10e with Android 12 since it almost works... the problem is, as I described earlier, that there must be at least one notification active for the app in order for the numeric badges to show. 

- The Pixel 4 and 5 emulators, which run Google's stock Android 11 and 12, behave worse in that numeric badges never show, only the dot. I have found a setting where I can turn the dot off but not replacing it with a number. Since Samsung supposedly ditched their  Android customisations long ago (I think that they nowadays use the stock Android more or less), my guess is that the emulators would work in the same way as my S10e if only ShortcutBadger would support the emulated devices. 

- Since supporting real devices is more important than 100% full support in the emulator, this boils down to just one conclusion: if the condition, that at least one notification has to be active for badges to show, could be eliminated, the badge plugin would work also on newer Android versions. At least that's what I think :) P.S. I just noticed some weird behaviour: if I read all the unread items in my app's inbox, this will decrement the badge number to 0, when the badge will be cleared. This did not happen, instead the badge remained at 1, which I realised was because of the still active notification. So an active notification is needed in order for the numeric badge to show but when the badge is cleared by the plugin, the number of active notifications "take over" the badge. So what needs to be done in order for the badge plugin to work as before is to disconnect completely Android's automatic handling of the app badge based on active notifications. I have no idea whether this can be done but hopefully this info is helpful in order to trace down a solution. I think that many applications would want the ability to control the badge without interference from the underlying OS.

**Reply by drtomato** [reply]
Apparently, it is possible to prevent badges when creating notifications. If it becomes possible to control badges completely programmatically, I think that the LocalNotifications plugin should disable badges by default (https://developer.android.com/training/notify-user/badges) and leave badge creation to the developer. 

However, I still haven't found a way to create badges without notifications. Users are also irritated: https://support.google.com/android/thread/13110750/can-you-have-notification-badges-without-notification-sounds-and-banners?hl=en

**Reply by drtomato** [reply]
Hi @robingenz,

More info: I went and bought myself the cheapest Android phone I could find to use for testing, namely the Alcatel 1 2021. It runs Android 11 Go, which is apparently an Android version suitable for really low-end devices. I though it would be interesting to see how things worked there, which was quick... it seems that there is no support for neither app icon dots nor numbers.

**Reply by robingenz** [reply]
@drtomato Thanks for all your updates. I'll get back to the topic soon. I'm just very busy right now.
